### PR TITLE
 http_server: serve license files 

### PR DESCRIPTION
**Pull request by hnez**
Some open source license require shipping the license text along with the built software.
To comply with this requirement we let yocto generate a directory with license texts inside of it and expose it via the web interface.

This PR is developed in tandem with linux-automation/meta-lxatac#163, which installs the license files to the image in the first place.

This is what this will look like to the user:

Legal Information Page

A new entry "Legal Information" was added to the web interface. Clicking on it reveals a list of installed packages.

---

License files

The rightmost column in the table contains links that lead to the license text provided by the package.
Clicking on it opens the plain text file in the browser.

### CODE REVIEW

**Comment by hnez** [comment]
I've assigned this to @jluebbe first to decide if this is what we want.
Deciding if this is implemented in a way we want is a separate task and can be done afterwards.

**Comment by hnez** [comment]
A stated in linux-automation/meta-lxatac#163 we also want to expose the license manifest file, because it includes software version information.
This will likely not need any modifications here, just a quick check if the file displays correctly and that the `tacd` does not stumble over it being a symlink (if we implement it that way).

**Comment by hnez** [comment]
> This will likely not need any modifications here, just a quick check if the file displays correctly and that the `tacd` does not stumble over it being a symlink (if we implement it that way).

Not quite how it turned out. Symlinking the manifest file would have required some hacks on the `meta-lxatac` side, so instead decided to "do it properly" and serve the manifest file and the license text files at different locations and use React code to tie them together.
The result can be seen in the screenshots above.
The manifest file is parsed in the web browser and is displayed as a table containing the package name, version and license. The license link is clickable and leads to the license text.

**Review by jluebbe — APPROVED** [review]

**Comment by hnez** [comment]
Hi @KarlK90, I know that web development is your secret passion.
Could you have a look at this PR from a Rust and Typescript perspective?

**Comment by KarlK90** [comment]
> Hi @KarlK90, I know that web development is your secret passion. Could you have a look at this PR from a Rust and Typescript perspective?

I'm sorry to inform you that this is not true, but I'll try my best. :wink:

File: web/src/Legal.tsx [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+
+function parse_mainifest(text: string) {
+  let packages: Package[] = [];
+
+  // The content of `text` looks something like this:
+  //
+  // PACKAGE NAME: tacd
+  // PACKAGE VERSION: 1.0.0
+  // RECIPE NAME: tacd
+  // LICENSE: GPL-2.0-or-later
```
Comment (KarlK90) [inline web/src/Legal.tsx:41]: How stable is the format generated by [`poky/meta/classes-recipe/license_image.bbclass`](https://github.com/yoctoproject/poky/blob/master/meta/classes-recipe/license_image.bbclass#L85-L88)? How hard would it be to have some kind of unit-test against a recent license file?

File: web/src/Legal.tsx [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+        {
+          id: "recipe_name",
+          header: "Recipe Name",
+          cell: (p) => p.recipe_name,
+        },
+        {
+          id: "license",
+          header: "License",
+          cell: (p) => (
+            <Link href={"/docs/legal/files/" + p.recipe_name}>{p.license}</Link>
```
Comment (KarlK90) [inline web/src/Legal.tsx:118]: I first thought that we could just set the mime type in the link on the front end, but we also generate a file listing. So you're approach is probably a good abstraction.

**Review by KarlK90 — CHANGES_REQUESTED** [review]
@hnez LGTM, but if possible I would like to see a unit-test for the license file.

**Comment by hnez** [comment]
> @hnez LGTM, but if possible I would like to see a unit-test for the license file.

Now that I have re-read this message I am not too sure if I actually implemented what you have asked for.
I've just pushed a few commits just now that add a web interface unit test for parsing the included demo_mode manifest file.

But that does not check against a recent _actual_ manifest file. That would be a bit more involved, because the manifest is generated in meta-lxatac, not in the tacd.

We _could_ add a test there, to check if the format did not change. But I think just keeping an eye on the yocto release notes for changes to the file format may be a better option. I would assume that a (breaking) change to the manifest file format would be treated as noteworthy enough to be listed in the release notes.

**Comment by SmithChart** [comment]
We could also add an integration test into `labgrid-lxatac` to check if the functionality on the actual device does still work. Or is the manifest processed during buildtime?

**Comment by SmithChart** [comment]
> The manifest is only processed at runtime, so that would be an option. The actual parsing does however take place in the browser and not on the TAC, making a full integration test complicated.

Indeed. I'll add it our test backlog anyway. So if we really want to we can add Selenium or what ever is the browser-test-engine-of-the-day to our dependencies.

**Comment by hnez** [comment]
Hi, it would be great to get this merged, since it is the only open PR before we can start preparing a new meta-lxatac release.

I am not quite sure if there is anything I need to do.

**Pull request merged**